### ISSUE TITLE
Improved Instrumentation of Models + Pre/Postprocessing

### ISSUE THREAD

**Issue by patrickhulce** [insight-platform/Savant#639]
First off, thanks for the amazing project! Super well documented and optimized for usability out of the box, bravo 👏 

One issue is that I'm struggling a bit to understand why an attribute model is taking so long (onnx model takes ~10ms in a notebook but ~100ms in this Savant pipeline), and I don't see immediate ways to instrument further from userland.

I think there are a couple things that could be done here though:

1. Automatically create spans for each processed object by default.
2. Automatically create spans for preprocessing and converter phases by default.
3. Expose the telemetry span object to the `preprocess_object_image` and `converter` modules so we can add our own detailed instrumentation.

I'm happy to contribute these changes if you're open to them! Just let me know :)

**Comment by bwsw** [comment]
Hello @patrickhulce 

I appreciate your interest in the project. Regarding your suggestions, let me clarify my vision on that:

Automatic spans for preprocessing and postporcessing functions are ok. If the user can access the parent span from their code to create custom child spans, it is what they usually need. If you implement that, we will accept the PR for sure.

> Automatically create spans for each processed object by default.

It is not ok because spans spend CPU, which can be excessive if one does not need them, e.g., in robotics. Please clarify your idea; maybe I understand it wrong.

> onnx model takes ~10ms in a notebook but ~100ms in this Savant pipeline

Please disclose what model you use and how you measure the processing time.

**Comment by patrickhulce** [comment]
Thanks for the reply!

> If you implement that, we will accept the PR for sure.

Great, will do then 👍 

> It is not ok because spans spend CPU, which can be excessive if one does not need them, e.g., in robotics. Please clarify your idea; maybe I understand it wrong.

Fair! Would you be open to option to enable it rather than by default? Or is there a different way to gain insight in OpenTelemetry as to the duration of each individual use of the model? i.e. I want to know in Jaeger if the span is large because it is slowly running on 1 object or whether it's running quickly on 100 objects. Anything that satisfies that goal would be fine with me.

> Please disclose what model you use and how you measure the processing time.

It's a custom faceswap model. In notebook I'm measuring the time it takes to crop on GPU (preprocessing step) and run a single forward pass of the model. In Savant I'm measuring the single span provided for the nvinfer pipeline element containing the model definition (which includes preprocessing + forward pass + converter time for all detected objects, right?). My goal here is to understand where that extra time is going. Is it indeed the model? Is it something else I need to tweak?

**Comment by bwsw** [comment]
> Fair! Would you be open to option to enable it rather than by default? Or is there a different way to gain insight in OpenTelemetry as to the duration of each individual use of the model? i.e. I want to know in Jaeger if the span is large because it is slowly running on 1 object or whether it's running quickly on 100 objects. Anything that satisfies that goal would be fine with me.

I doubt it is a straightforward thing. Because GStreamer is a parallel system, sometimes you cannot make sure that data is under processing or just idling somewhere because of a bottleneck. Thus, it is only usable (according to my thoughts) to profile the code that is immediately responsible for the processing. As a result, you can see spans like:

```
|-----------------------------------------------|
 ======IDLE=====|---------------|======IDLE=====
```

OpenTelemetry tracing represents a flow for a particular frame, but you want to get stats across multiple frames. It can be extracted from Jaeger but is not accompanied by tooling.

**To sum up.** You can introduce the span access API feature for pre- and post-processing, but it must not create spans per object; it would be better to allow users to do that with API if they need it. But measuring how long it takes a whole pre- or post-processing is a good thing.

**Comment by bwsw** [comment]
You also can find the yolov8_seg demo useful for you, as it introduces a heavy model and complex mask operations. The pipeline has a sample of postprocessing on GPU with CuPy:

https://github.com/insight-platform/Savant/tree/develop/samples/yolov8_seg/module
